core/types/transaction_conditional.go

```diff
@@ -0,0 +1,130 @@
+)
+
+// KnownAccounts represents a set of KnownAccounts
+type KnownAccounts map[common.Address]KnownAccount
+
+// KnownAccount allows for a user to express their preference of a known
+// prestate at a particular account. Only one of the storage root or
+// storage slots is allowed to be set. If the storage root is set, then
+// the user prefers their transaction to only be included in a block if
+// the account's storage root matches. If the storage slots are set,
+// then the user prefers their transaction to only be included if the
+// particular storage slot values from state match.
+type KnownAccount struct {
+	StorageRoot  *common.Hash `rlp:"nil"`
+	StorageSlots map[common.Hash]common.Hash
+}
+
+// UnmarshalJSON will parse the JSON bytes into a KnownAccount struct.
+func (ka *KnownAccount) UnmarshalJSON(data []byte) error {
+	var hash common.Hash
+	if err := json.Unmarshal(data, &hash); err == nil {
+		ka.StorageRoot = &hash
+		ka.StorageSlots = make(map[common.Hash]common.Hash)
+		return nil
+	}
+
+	var mapping map[common.Hash]common.Hash
+	if err := json.Unmarshal(data, &mapping); err != nil {
+		return err
+	}
+	ka.StorageSlots = mapping
+	return nil
+}
+
+// MarshalJSON will serialize the KnownAccount into JSON bytes.
+func (ka *KnownAccount) MarshalJSON() ([]byte, error) {
+	if ka.StorageRoot != nil {
+		return json.Marshal(ka.StorageRoot)
+	}
+	return json.Marshal(ka.StorageSlots)
+}
+
+// Root will return the storage root and true when the user prefers
+// execution against an account's storage root, otherwise it will
+// return false.
+func (ka *KnownAccount) Root() (common.Hash, bool) {
+	if ka.StorageRoot == nil {
+		return common.Hash{}, false
+	}
+	return *ka.StorageRoot, true
+}
+
+// Slots will return the storage slots and true when the user prefers
+// execution against an account's particular storage slots, StorageRoot == nil,
+// otherwise it will return false.
+func (ka *KnownAccount) Slots() (map[common.Hash]common.Hash, bool) {
+	if ka.StorageRoot != nil {
+		return ka.StorageSlots, false
+	}
+	return ka.StorageSlots, true
+}
+
+//go:generate go run github.com/fjl/gencodec -type TransactionConditional -field-override transactionConditionalMarshalling -out gen_transaction_conditional_json.go
+
+// TransactionConditional represents the preconditions that determine the
+// inclusion of the transaction, enforced out-of-protocol by the sequencer.
+type TransactionConditional struct {
+	// KnownAccounts represents account prestate conditions
+	KnownAccounts KnownAccounts `json:"knownAccounts"`
+
+	// Header state conditionals (inclusive ranges)
+	BlockNumberMin *big.Int `json:"blockNumberMin,omitempty"`
+	BlockNumberMax *big.Int `json:"blockNumberMax,omitempty"`
+	TimestampMin   *uint64  `json:"timestampMin,omitempty"`
+	TimestampMax   *uint64  `json:"timestampMax,omitempty"`
+
+	// Internal fields used for metrics and inclusion tracking
+	SubmissionTime time.Time    `json:"-"`
```

Comment:
This is decorated by the proxy? There is already a time property on the transaction:
- https://github.com/ethereum-optimism/op-geth/blob/60038121c7571a59875ff9ed7679c48c9f73405d/core/types/transaction.go#L58
- https://github.com/ethereum-optimism/op-geth/blob/60038121c7571a59875ff9ed7679c48c9f73405d/core/types/transaction.go#L561

Reply:
ah no it's not, didn't realize this was available! I'll edit to use this instead.

Reply:
Looking at the docstring and usage. It looks like this is only used in tests

Going to update the docstring as well and set the time in the API handler

Reply:
It is used in the block building to break ties on same fee - https://github.com/ethereum/go-ethereum/blob/d71831255da7ecb3817c9e14c9142fe9d4441d3b/miner/ordering.go#L67

Reply:
ahh okay I see, thanks